### ISSUE TITLE
Bug: illegal moves with version 0.3.2

### ISSUE THREAD

**Issue by rwbc** [megalodon-chess/megalodon#50]
Both times Megalodon 0.3.2 moved through an opponents piece ;-)
The second position also involves a double check, which might be fixed minutes ago...

```
[Event "Test@CuteCGdv"]
[Site "RWBC-CAPPUCCINO Win7U64 Q8200 2.33Ghz + Nvidia GT 710"]
[Date "2021.04.01"]
[Round "7"]
[White "Megalodon_032-64"]
[Black "NSVChess_014"]
[Result "0-1"]
[ECO "B30"]
[GameDuration "00:01:01"]
[GameEndTime "2021-04-01T18:45:21.621 Mitteleuropäische Sommerzeit"]
[GameStartTime "2021-04-01T18:44:20.258 Mitteleuropäische Sommerzeit"]
[Opening "Sicilian"]
[PlyCount "22"]
[Termination "illegal move"]
[TimeControl "40/120"]
[Variation "Nimzovich-Rossolimo attack (without ...d6)"]

1. e4 {book} c5 {book}
2. Nf3 {book} Nc6 {book}
3. Bb5 {book} Nd4 {book}
4. Nxd4 {+1.30/6 1.64} cxd4 {-0.12/5 2.30}
5. Bd3 {-1.05/5 1.68} d5 {+0.07/5 5.17}
6. Bb5+ {-0.02/5 1.73} Bd7 {+0.28/5 0.080}
7. Bxd7+ {+1.42/6 1.79} Qxd7 {+0.24/5 0.18}
8. exd5 {+1.38/6 1.84} Qxd5 {+0.25/5 5.06}
9. O-O {+2.29/6 1.90} e5 {+0.32/5 10.40}
10. f4 {+0.02/5 1.97} exf4 {+1.27/5 5.65}
11. Rxf4 {+2.51/6 2.04} Bd6 {+0.27/5 15.72}
{White makes an illegal move: f4a4}
0-1
```

```
[Event "Test@CuteCGdv"]
[Site "RWBC-CAPPUCCINO Win7U64 Q8200 2.33Ghz + Nvidia GT 710"]
[Date "2021.04.01"]
[Round "8"]
[White "Megalodon_032-64"]
[Black "NSVChess_014"]
[Result "0-1"]
[ECO "A00"]
[GameDuration "00:02:06"]
[GameEndTime "2021-04-01T18:50:52.536 Mitteleuropäische Sommerzeit"]
[GameStartTime "2021-04-01T18:48:45.924 Mitteleuropäische Sommerzeit"]
[Opening "Benko's Opening"]
[PlyCount "40"]
[Termination "illegal move"]
[TimeControl "40/120"]

1. g3 {book} d5 {book}
2. d3 {book} c5 {book}
3. Bg2 {book} Nc6 {book}
4. c4 {-1.32/5 1.64} dxc4 {+1.33/5 3.59}
5. dxc4 {-1.25/5 1.68} Qxd1+ {+9.18/5 4.34}
6. Kxd1 {+1.20/7 1.73} Ne5 {+0.21/5 3.92}
7. Nd2 {-1.37/5 1.79} Ng4 {-0.14/5 3.12}
8. Ke1 {+1.27/6 1.83} e5 {+0.07/5 4.37}
9. e4 {+1.35/6 1.91} Bd6 {-0.02/5 3.22}
10. Ngf3 {+1.44/6 2.01} f5 {-0.15/5 5.68}
11. Kf1 {-0.96/5 2.05} N8f6 {+0.14/5 3.13}
12. exf5 {+1.30/6 3.84} Bxf5 {+0.39/5 5.50}
13. Ne1 {-1.27/5 2.19} e4 {+0.60/5 5.00}
14. Nc2 {+1.27/6 2.29} e3 {+0.54/5 14.88}
15. Bxb7 {+2.60/6 2.39} exd2 {+3.02/5 5.03}
16. Bxa8 {+1.37/6 2.52} d1=Q+ {+5.92/5 2.65}
17. Ne1 {-5.68/6 2.68} O-O {+6.21/5 14.66}
18. Bb7 {-7.68/5 2.84} Nxf2 {+7.08/4 1.45}
19. Kxf2 {-4.06/6 3.04} Ne4+ {+3.73/4 1.74}
20. Kf1 {-4.41/6 3.28} Bh3+ {+3.57/4 0.98}
{White makes an illegal move: b7e4}
0-1
```

**Comment by ArjunSahlot** [comment]
This seems to be a push function bug. Since when I used the fen to get legal moves, all of them were legal.

**Comment by ArjunSahlot** [comment]
And yes, the double check bug was recently fixed.

**Comment by HuangPatrick16777216** [comment]
Thank you for your bug report. The first bug was a push bug, as stated by @ArjunSahlot. It has been fixed in e331d45.
